File: trunk/src/test/slc-dev/isink.c
/* 
 * File:	 usb-test.c
 * Author:   Robinson Mittmann (bobmittmann@gmail.com)
 * Target:
 * Comment:
 * Copyright(C) 2011 Bob Mittmann. All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

#include "board.h"
#include "isink.h"
#include <sys/dcclog.h>

#define FREQ_1MHZ 1000000

void isink_start(unsigned int mode, unsigned int pre, unsigned int pulse)
{
	struct stm32f_tim * tim = STM32_TIM4;

	if (trig_mode == TRIG_ADDR) {
		stm32_gpio_mode(SINK1, INPUT, 0);
		stm32_gpio_mode(SINK2, INPUT, 0);
		stm32_gpio_mode(SINK3, INPUT, 0);
		return;
	}

	tim->arr = pre;

	switch (mode) {
	case 0:
		stm32_gpio_mode(SINK1, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK2, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK3, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		tim->ccr2 = pulse;
		tim->ccr3 = pre;
		tim->ccr4 = pre;
		break;
	case 1:
		stm32_gpio_mode(SINK1, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK2, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK3, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		tim->ccr2 = pre;
		tim->ccr3 = pulse;
		tim->ccr4 = pre;
		break;
	case 2:
		stm32_gpio_mode(SINK1, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK2, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK3, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		tim->ccr2 = pre;
		tim->ccr3 = pre;
		tim->ccr4 = pulse;
		break;
	case 3:
		stm32_gpio_mode(SINK1, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK2, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK3, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		tim->ccr2 = pulse;
		tim->ccr3 = pulse;
		tim->ccr4 = pre;
		break;
	case 4:
		stm32_gpio_mode(SINK1, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK2, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK3, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		tim->ccr2 = pre;
		tim->ccr3 = pulse;
		tim->ccr4 = pulse;
		break;
	case 5:
		stm32_gpio_mode(SINK1, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK2, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK3, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		tim->ccr2 = pulse;
		tim->ccr3 = pre;
		tim->ccr4 = pulse;
		break;
	case 6:
		stm32_gpio_mode(SINK1, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK2, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK3, OUTPUT, PUSH_PULL | SPEED_HIGH);
		tim->ccr2 = pre;
		tim->ccr3 = pulse;
		break;
	case 7:
		stm32_gpio_mode(SINK1, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK2, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK3, OUTPUT, PUSH_PULL | SPEED_HIGH);
		tim->ccr2 = pulse;
		tim->ccr3 = pre;
		break;
	case 8:
		stm32_gpio_mode(SINK1, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK2, OUTPUT, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK3, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		tim->ccr2 = pre;
		tim->ccr4 = pulse;
		break;
	case 9:
		stm32_gpio_mode(SINK1, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK2, OUTPUT, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK3, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		tim->ccr2 = pulse;
		tim->ccr4 = pre;
		break;
	case 10:
		stm32_gpio_mode(SINK1, OUTPUT, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK2, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK3, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		tim->ccr3 = pre;
		tim->ccr4 = pulse;
		break;
	case 11:
		stm32_gpio_mode(SINK1, OUTPUT, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK2, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		stm32_gpio_mode(SINK3, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
		tim->ccr3 = pulse;
		tim->ccr4 = pre;
		break;
	}

 	tim->cnt = pulse + pre;
	/* trigger the timer */
	tim->cr1 = TIM_CMS_EDGE | TIM_DIR_DOWN | TIM_OPM | TIM_URS | TIM_CEN; 
}

void isink_stop(void)
{
	stm32_gpio_mode(SINK1, INPUT, 0);
	stm32_gpio_mode(SINK2, INPUT, 0);
	stm32_gpio_mode(SINK3, INPUT, 0);
	stm32_gpio_mode(SINK4, INPUT, 0);
	stm32_gpio_clr(SINK1);
	stm32_gpio_clr(SINK2);
	stm32_gpio_clr(SINK3);
	stm32_gpio_clr(SINK4);
	stm32_gpio_mode(SINK1, OUTPUT, PUSH_PULL | SPEED_MED);
	stm32_gpio_mode(SINK2, OUTPUT, PUSH_PULL | SPEED_MED);
	stm32_gpio_mode(SINK3, OUTPUT, PUSH_PULL | SPEED_MED);
	stm32_gpio_mode(SINK4, OUTPUT, PUSH_PULL | SPEED_MED);
}

void isink_pulse(unsigned int pre, unsigned int pulse)
{
	struct stm32f_tim * tim = STM32_TIM4;

	stm32_gpio_mode(SINK1, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
	stm32_gpio_mode(SINK2, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
	stm32_gpio_mode(SINK3, OUTPUT, PUSH_PULL | SPEED_HIGH);

	tim->ccr2 = pulse;
	tim->ccr3 = pre;
	tim->arr = pre;
 	tim->cnt = pulse + pre;
	tim->cr1 = TIM_CMS_EDGE | TIM_DIR_DOWN | TIM_OPM | TIM_URS | TIM_CEN; 
}

void isink_pulse_dual(unsigned int pre, unsigned int pulse)
{
	struct stm32f_tim * tim = STM32_TIM4;

	stm32_gpio_mode(SINK1, ALT_FUNC, PUSH_PULL | SPEED_HIGH);
	stm32_gpio_mode(SINK2, OUTPUT, PUSH_PULL | SPEED_HIGH);
	stm32_gpio_mode(SINK3, ALT_FUNC, PUSH_PULL | SPEED_HIGH);

	tim->ccr2 = pre;
	tim->ccr4 = pulse;
	tim->arr = pre;
 	tim->cnt = pulse + pre;
	tim->cr1 = TIM_CMS_EDGE | TIM_DIR_DOWN | TIM_OPM | TIM_URS | TIM_CEN; 
}

void isink_init(void)
{
	struct stm32f_dac * dac = STM32_DAC;
	struct stm32f_tim * tim = STM32_TIM4;
	uint32_t div;

	stm32_gpio_clr(SINK1);
	stm32_gpio_clr(SINK2);
	stm32_gpio_clr(SINK3);
	stm32_gpio_clr(SINK4);
	stm32_gpio_mode(SINK1, OUTPUT, PUSH_PULL | SPEED_HIGH);
	stm32_gpio_mode(SINK2, OUTPUT, PUSH_PULL | SPEED_HIGH);
	stm32_gpio_mode(SINK3, OUTPUT, PUSH_PULL | SPEED_HIGH);
	stm32_gpio_mode(SINK4, OUTPUT, PUSH_PULL | SPEED_HIGH);
	stm32_gpio_af(SINK1, GPIO_AF2);
	stm32_gpio_af(SINK2, GPIO_AF2);
	stm32_gpio_af(SINK3, GPIO_AF2);
	stm32_gpio_af(SINK4, GPIO_AF2);

	/* Timer clock enable */
	stm32_clk_enable(STM32_RCC, STM32_CLK_TIM4);
	
	/* get the total divisior */
	div = (stm32f_tim1_hz + (FREQ_1MHZ / 2)) / FREQ_1MHZ;
	/* Timer configuration */
	tim->cr1 = 0;
	tim->psc = div - 1;
	tim->arr = 100;
	tim->ccmr1 = TIM_OC1M_PWM_MODE2 | TIM_OC2M_PWM_MODE2;
	tim->ccmr2 = TIM_OC3M_PWM_MODE2 | TIM_OC4M_PWM_MODE2;
	tim->ccer = TIM_CC1E | TIM_CC2E | TIM_CC3E | TIM_CC4E;
	//	TIM_CC1P | TIM_CC2P | TIM_CC3P | TIM_CC4P;
	tim->ccr1 = 0;
	tim->ccr2 = 0;
	tim->ccr3 = 0;
	tim->ccr4 = 0;


	/* I/O pins config */
	stm32_gpio_mode(IRATE, ANALOG, 0);
	stm32_gpio_clr(IRATE);

	/* DAC clock enable */
	stm32_clk_enable(STM32_RCC, STM32_CLK_DAC);

	/* DAC disable */
	dac->cr = 0;
	/* DAC configure */
	dac->cr = DAC_EN2 | DAC_EN1;
	/* DAC channel 2 initial value */
	dac->dhr12r2 = 0;
	/* DAC channel 1 initial value */
	dac->dhr12r1 = 0;
}

void isink_slewrate_set(unsigned int rate)
{
	struct stm32f_dac * dac = STM32_DAC;
	unsigned int dac_val;

	if (rate < SLEWRATE_MIN)
		rate = SLEWRATE_MIN;
	else if (rate >= SLEWRATE_MAX)
		rate = SLEWRATE_MAX;

	dac_val = (4095 * (rate - SLEWRATE_MIN)) / (SLEWRATE_MAX - SLEWRATE_MIN); 
//	DCC_LOG2(LOG_TRACE, "rate=%dmA/us DAC=%d", rate, dac_val);

	dac->dhr12r2 = dac_val;
}

File: src/slcdev/sim/app/isink.py
#!/usr/bin/env python

#
# copyright(c) 2012 robinson mittmann. all rights reserved.
# 
# this file is part of the yard-ice.
#
# this library is free software; you can redistribute it and/or
# modify it under the terms of the gnu lesser general public
# license as published by the free software foundation; either
# version 3.0 of the license, or (at your option) any later version.
# 
# this library is distributed in the hope that it will be useful,
# but without any warranty; without even the implied warranty of
# merchantability or fitness for a particular purpose.  see the gnu
# lesser general public license for more details.
# 
# you can receive a copy of the gnu lesser general public license from 
# http://www.gnu.org/

#
#    file: isink.py
# comment: current sink waveform 
#  author: robinson mittmann <bobmittmann@gmail.com>
# 

#!/usr/bin/env python

def isink(mode):
	
	r40 = 66500.0 # pwm output - SINK1
	r30 = 44200.0 # pwm output - SINK2
	r35 = 39200.0 # pwm output - SINK3
	r39 = 24000.0 # pwm output - SINK4
	r32 =  6190.0 # GND

	r = [ r40, r30, r35, r39 ];

	i = 0
	# gnd connected equivalent resisor
	ra = r32
	while i < 4:
		if mode[i] == 0:
			ra = (ra * r[i]) / (ra + r[i])
		i = i + 1

	i = 0
	#vcc connected equivalent resistor
	rb = 10e9 
	while i < 4:
		if mode[i] == 1:
			rb = (rb * r[i]) / (rb + r[i])
		i = i + 1

	v = 3.3 * ra / (ra + rb)
	i = v / 4.99;

#	if (rb > 10e6):
#		print("RA={0:5.2f}K RB=~     ".format(ra / 1000, rb / 1000));
#	else:
#		print("RA={0:5.2f}K RB={1:5.2f}K "\
#			  "i={2:.0f}mA".format(ra / 1000, rb / 1000, i * 1000))

	return int(round(i * 1000))

#	print("RA={0:.2f}K, RB={1:.2f}K, i={2:.0f}mA"\
#		  .format(ra / 1000, rb/ 1000), i);

def print_io(x):
	j = 0
	while j < 4:
		if (x[j][0] == 1):
			print("\t\tstm32_gpio_mode(IO_SINK{0:d}, ALT_FUNC, "\
				  "PUSH_PULL | SPEED_HIGH);".format(j + 1))
		if (x[j][0] == 2): 
			print("\t\tstm32_gpio_mode(IO_SINK{0:d}, ALT_FUNC, "\
				  "OPEN_DRAIN | SPEED_HIGH);".format(j + 1))
		if (x[j][0] == 0):
			print("\t\tstm32_gpio_mode(IO_SINK{0:d}, OUTPUT, "\
				  "PUSH_PULL | SPEED_HIGH);".format(j + 1))
		j = j + 1

	j = 0
	while j < 4:
		if x[j][1] == 0:
			print("\t\ts{:d} = -1;".format(j + 1))
		else:
			print("\t\ts{:d} = 0;".format(j + 1))
		j = j + 1

def isink_dump(lst):
	nm = [ 'GND', 'VCC', 'NC ']

	r40 = 66500.0 # pwm output - SINK1
	r30 = 44200.0 # pwm output - SINK2
	r35 = 39200.0 # pwm output - SINK3
	r39 = 24000.0 # pwm output - SINK4
	r32 =  6190.0 # GND

	print('R40 = {:6.3f}K'.format(r40 / 1000))
	print('R30 = {:6.3f}K'.format(r30 / 1000))
	print('R35 = {:6.3f}K'.format(r35 / 1000))
	print('R39 = {:6.3f}K'.format(r39 / 1000))
	print('R32 = {:6.3f}K'.format(r32 / 1000))
	print

	print('| Cnt | i(mA) | v(mv) | R40 | R30 | R35 | R39 |')
	print('+-----+-------+-------+-----+-----+-----+-----+')
	j = 0
	for e in lst:
		s = "| {:3d} | {:5d} | {:5d} | " .format(j, e[0], e[0] * 5)
		k = 1
		while k < 5:
			s += nm[e[k]] + " | "
			k = k + 1
		print(s)
		j = j + 1

	print('+-----+-------+-------+-----+-----+-----+-----+')

def v(x, y):
	return (x == y) or ((x > 0) and (y == 0))

def main():

	i_lst = []
	nm = [ 'GND ', 'VCC ', 'NC  ']

	w = 0
	while w < 3:
		z = 0
		while z < 3:
			y = 0
			while y < 3:
				x = 0
				while x < 3:
					i = isink([x, y, z, w]);
					if (i > 0):
						i_lst.append([i, x, y, z, w])
					x = x + 1
				y = y + 1
			z = z + 1
		w = w + 1

	i_lst.sort()

#	isink_dump(i_lst)
#	return

	q_lst = []
	for n in i_lst:
		for m in i_lst:
			i0 = n[0]
			w0 = n[1]
			x0 = n[2]
			y0 = n[3]
			z0 = n[4]
			i1 = m[0]
			w1 = m[1]
			x1 = m[2]
			y1 = m[3]
			z1 = m[4]
			if (i0 >= 1.75 * i1) and v(w0, w1) and v(x0, x1) \
					  and v(y0, y1) and v(z0, z1):
				q_lst.append([n, m])

	p_lst = q_lst
	for x in q_lst:
		for y in q_lst:
			d0 = abs(x[0][0] - y[0][0])
			d1 = abs(x[1][0] - y[1][0])
			ref0 = x[0][0] * 0.005
			ref1 = x[1][0] * 0.001
			if (d1 < ref1) and (d0 < ref0):
				p_lst.remove(y)
		

	print("static void isink_io_cfg(unsigned int mode)")
	print("{")
	print("\tint32_t s1;")
	print("\tint32_t s2;")
	print("\tint32_t s3;")
	print("\tint32_t s4;")
	print("")
	print("\tswitch (mode) {")
	print("")

	j = 0
	for n in p_lst:
		i0 = n[0][0]
		w0 = n[0][1]
		x0 = n[0][2]
		y0 = n[0][3]
		z0 = n[0][4]
		i1 = n[1][0]
		w1 = n[1][1]
		x1 = n[1][2]
		y1 = n[1][3]
		z1 = n[1][4]
		print("\tcase {0:d}:".format(j))
		j = j + 1
		s = "\t/* ("
		s += "{:d} {:d})[mA] ({:d} {:d})[mV]".format(i0, i1, i0 * 5, i1 * 5)
		s += " [ " + nm[w0] + nm[x0] + nm[y0] + nm[z0] + "]" 
		s += " [ " + nm[w1] + nm[x1] + nm[y1] + nm[z1] + "] */"
		print(s)

		print_io([[w0, w1], [x0, x1], [y0, y1], [z0, z1]])

		print("\t\tbreak;")
		print("")
	print("\t}")
	print("")

	print("\tisink_drv.s1 = s1;")
	print("\tisink_drv.s2 = s2;")
	print("\tisink_drv.s3 = s3;")
	print("\tisink_drv.s4 = s4;")
	print("}")
	print("")

	print("/*")
	print("\t| mode | i0(mA) | i1(mA) |")
	print("\t+------+--------+--------+")
	j = 0
	for n in p_lst:
		i0 = n[0][0]
		i1 = n[1][0]
		print("\t| {:4d} | {:6d} | {:6d} |" .format(j, i0, i1))
		j = j + 1

	print("\t+------+--------+--------+")
	print("*/");
	print("");

if __name__ == '__main__':
	main()

